File: 102.binary-tree-level-order-traversal.py
#
# @lc app=leetcode id=102 lang=python3
#
# [102] Binary Tree Level Order Traversal
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
# from typing import *
from queue import Queue
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        # if not root.left and not root.left
        res = []
        current_level_values = []
        current_queue = []
        next_queue = []
        current_queue.append(root)

        while current_queue:
            node = current_queue.pop(0)
            if node.left:
                next_queue.append(node.left)
            if node.right:
                next_queue.append(node.right)
            current_level_values.append(node.val)
            if not current_queue:
                res.append(current_level_values)
                current_level_values = []
                if next_queue:
                    current_queue = next_queue
                    next_queue = []
        
        return res

# @lc code=end

File: 102.binary-tree-level-order-traversal.go
/*
 * @lc app=leetcode id=102 lang=golang
 *
 * [102] Binary Tree Level Order Traversal
 */
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// @lc code=start
/**
* Definition for a binary tree node.
* type TreeNode struct {
*     Val int
*     Left *TreeNode
*     Right *TreeNode
* }
 */
//  keep in sync with 104.maximum-depth-of-binary-tree
func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	curLayerNode := []*TreeNode{root}
	result := [][]int{}
	for len(curLayerNode) != 0 {
		curLayerValue := []int{}
		nextLayerNode := []*TreeNode{}
		for _, node := range curLayerNode {
			curLayerValue = append(curLayerValue, node.Val)
			if node.Left != nil {
				nextLayerNode = append(nextLayerNode, node.Left)
			}
			if node.Right != nil {
				nextLayerNode = append(nextLayerNode, node.Right)
			}
		}
		result = append(result, curLayerValue)
		curLayerNode = nextLayerNode
	}
	return result
}

// @lc code=end
